From fd9db2bd6165314939fa076c59bc4387e2bb50ba Mon Sep 17 00:00:00 2001
From: Dipen Dudhat <dipen.dudhat@freescale.com>
Date: Thu, 8 Oct 2009 16:51:31 +0530
Subject: [PATCH] eSDHC controller on P2020RDB has bug for data transfer using DMA.
For that driver has configured to do data transfer using CPU.


Signed-off-by: Dipen Dudhat <dipen.dudhat@freescale.com>
---
 arch/powerpc/boot/dts/p2020rdb.dts |    1 +
 drivers/mmc/host/sdhci-of.c        |    3 +++
 drivers/mmc/host/sdhci.c           |    3 ++-
 3 files changed, 6 insertions(+), 1 deletions(-)

diff --git a/arch/powerpc/boot/dts/p2020rdb.dts b/arch/powerpc/boot/dts/p2020rdb.dts
index da4cb0d..2ec3057 100644
--- a/arch/powerpc/boot/dts/p2020rdb.dts
+++ b/arch/powerpc/boot/dts/p2020rdb.dts
@@ -480,6 +480,7 @@
 			reg = <0x2e000 0x1000>;
 			interrupts = <72 0x2>;
 			interrupt-parent = <&mpic>;
+			fsl,sdhci-dma-broken;
 			/* Filled in by U-Boot */
 			clock-frequency = <0>;
 		};
diff --git a/drivers/mmc/host/sdhci-of.c b/drivers/mmc/host/sdhci-of.c
index 01ab916..5c56356 100644
--- a/drivers/mmc/host/sdhci-of.c
+++ b/drivers/mmc/host/sdhci-of.c
@@ -270,6 +270,9 @@ static int __devinit sdhci_of_probe(struct of_device *ofdev,
 	if (sdhci_of_wp_inverted(np))
 		host->quirks |= SDHCI_QUIRK_INVERTED_WRITE_PROTECT;
 
+	if (of_get_property(np, "fsl,sdhci-dma-broken", NULL))
+		host->quirks |= SDHCI_QUIRK_BROKEN_DMA;
+
 	clk = of_get_property(np, "clock-frequency", &size);
 	if (clk && size == sizeof(*clk) && *clk)
 		of_host->clock = *clk;
diff --git a/drivers/mmc/host/sdhci.c b/drivers/mmc/host/sdhci.c
index f5c39d8..8d9b30f 100644
--- a/drivers/mmc/host/sdhci.c
+++ b/drivers/mmc/host/sdhci.c
@@ -179,7 +179,8 @@ static void sdhci_init(struct sdhci_host *host)
 	sdhci_reset(host, SDHCI_RESET_ALL);
 
 	/* Enable cache snooping */
-	sdhci_writel(host, SDHCI_CACHE_SNOOP, SDHCI_HOST_DMA_CONTROL);
+	if (!(host->quirks & SDHCI_QUIRK_BROKEN_DMA))
+		sdhci_writel(host, SDHCI_CACHE_SNOOP, SDHCI_HOST_DMA_CONTROL);
 
 	sdhci_clear_set_irqs(host, SDHCI_INT_ALL_MASK,
 		SDHCI_INT_BUS_POWER | SDHCI_INT_DATA_END_BIT |
-- 
1.5.6.3

